### PR TITLE
Factor out numerical logic

### PR DESCRIPTION
**Pull request by soulitzer**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #55871 Refactor fast gradcheck
* #55699 Enable fixed fast_mode for complex
* #55692 Refactor C->C to C->R twice
* #55656 Clean up raise exception logic
* #55237 Enable fast gradcheck for real inputs and outputs
* #54480 Implement faster gradcheck but not enabled for most things
* **#54479 Factor out numerical logic**
* #54378 Refactor get numerical jacobian to calculate wrt all outputs at once

This change is similar to #54049 in that it helps us factor out some code that can be used in both fast and slow versions of gradcheck.
 - `compute_gradient` and `compute_numerical_jacobian_cols` have  fewer responsibilities:
   - compute_numerical_jacobian_cols essentially only handles the complexity of complex derivatives
   - compute_gradient handles only finite differencing (and doesn't worry about different layouts and indexing into the input tensor)
  - we have two stages again where we first compute the columns separately, then combine them

Differential Revision: [D27728727](https://our.internmc.facebook.com/intern/diff/D27728727)

### CODE REVIEW

File: torch/autograd/gradcheck.py [hunk 1]
```diff
@@ -252,8 +275,8 @@ def check_jacobians_equal(j1, j2, atol):
     return True
 
 
-def combine_jacobian_rows(jacobians_rows, inputs, output):
-    out_jacobians = allocate_jacobians_with_inputs(inputs, output.numel())
+def combine_jacobian_rows(jacobians_rows, inputs, dim):
```
Comment (zou3519) [inline torch/autograd/gradcheck.py:278]: nit: dim -> numel or "output_numel" might make more sense? `dim` is an overloaded word, it can mean "the number of dimensions in your tensor, e.g. tensor.dim()" or "the size of dimension X in your tensor, e.g. tensor.shape[X]"

File: torch/autograd/gradcheck.py [hunk 2]
```diff
@@ -141,7 +141,70 @@ def get_numerical_jacobian(fn, inputs, outputs=None, target=None, eps=1e-3,
...
+                jacobians_cols.append(torch.real(grad_out.conjugate() * ds_dx))
+            else:                   # R -> R
+                # skip if grad_out is complex but output is real
+                if not isinstance(grad_out, complex):
+                    jacobians_cols.append(ds_dx * grad_out)
+                else:
+                    jacobians_cols.append(None)
+
+
+def combine_jacobian_cols(jacobians_cols, outputs, input, dim):
```
Comment (zou3519) [inline torch/autograd/gradcheck.py:203]: nit: dim -> numel

Comment (zou3519) [inline torch/autograd/gradcheck.py:203]: Type hints would really help here (and in some of the other functions). E.g. it looks like `jacobian_cols` is always `Dict[int, List[Optional[Tensor]]` and this function returns a... `List[List[Tensor]]` ?

File: torch/autograd/gradcheck.py [hunk 3]
```diff
@@ -141,7 +141,70 @@ def get_numerical_jacobian(fn, inputs, outputs=None, target=None, eps=1e-3,
...
-        jacobians += [get_numerical_jacobian_helper(fn, inp, inp_idx, inputs, outputs, eps, grad_out)]
+        jacobians += [get_numerical_jacobian_wrt_specific_input(fn, inp, inp_idx, inputs, outputs, eps, eps, grad_out)]
+    return jacobians
+
+
+def compute_numerical_gradient(fn, entry, v, norm_v, nbhd_checks_fn):
+    # Performs finite differencing by perturbing `entry` in-place by `v` and
+    # returns the gradient of each of the outputs wrt to x at idx.
+    # we currently assume that the norm of delta equals eps
+    # assert(v == norm_v or v == (norm_v * 1j))
```
Comment (zou3519) [inline torch/autograd/gradcheck.py:152]: Where are `delta` and `eps` coming from? Should the assert be uncommented?

File: torch/autograd/gradcheck.py [hunk 4]
```diff
@@ -141,7 +141,70 @@ def get_numerical_jacobian(fn, inputs, outputs=None, target=None, eps=1e-3,
...
+
+    v = v.reshape(entry.shape) if isinstance(v, torch.Tensor) else v
```
Comment (zou3519) [inline torch/autograd/gradcheck.py:154]: `v` looks like it is always a Python number before this PR (I could be wrong), why did that change?

File: torch/autograd/gradcheck.py [hunk 5]
```diff
@@ -141,7 +141,70 @@ def get_numerical_jacobian(fn, inputs, outputs=None, target=None, eps=1e-3,
         target = inputs
     inp_indices = [i for i, a in enumerate(target) if is_tensor_like(a) and a.requires_grad]
     for i, (inp, inp_idx) in enumerate(zip(iter_tensors(target, True), inp_indices)):
-        jacobians += [get_numerical_jacobian_helper(fn, inp, inp_idx, inputs, outputs, eps, grad_out)]
+        jacobians += [get_numerical_jacobian_wrt_specific_input(fn, inp, inp_idx, inputs, outputs, eps, eps, grad_out)]
+    return jacobians
+
+
+def compute_numerical_gradient(fn, entry, v, norm_v, nbhd_checks_fn):
```
Comment (zou3519) [inline torch/autograd/gradcheck.py:148]: Would it be simpler if we computed `norm_v` inside this function instead of passing it in as an argument? Depending on what `v` can be this is not too computationally expensive

File: torch/autograd/gradcheck.py [hunk 6]
```diff
@@ -180,68 +243,28 @@ def check_outputs_same_dtype_and_shape_in_neighborhood(output1, output2, idx, de
...
+                        for i, a in enumerate(_as_tuple(inputs)))
+            return tuple(a.clone() for a in _as_tuple(fn(*inp)))
+
+        input_to_perturb = x[idx]
+        nbhd_checks_fn = functools.partial(check_outputs_same_dtype_and_shape_in_neighborhood, idx=idx, delta=delta)
+
+        def jvp_fn(delta):
+            return compute_numerical_gradient(wrapped_fn, input_to_perturb, delta, eps, nbhd_checks_fn)
+        jacobian_cols[d_idx] = []
+        compute_numerical_jacobian_cols(jacobian_cols[d_idx], delta, jvp_fn, x.is_complex(), grad_out)
```
Comment (zou3519) [inline torch/autograd/gradcheck.py:265]: Mutation is difficult to reason about when there are a lot of variables flying around. Could we have something like the following?
```
jacobain_cols[d_idx] = compute_numerical_jacobian_cols(delta, jvp_fn, x.is_complex(), grad_out)
```

**Review by zou3519 — COMMENTED** [review]
some initial comments/questions

File: torch/autograd/gradcheck.py (hunk 3 above)
Comment (soulitzer) [reply torch/autograd/gradcheck.py:152]: Since we are computing norm_v inside the function now, lets just remove this assert entirely

File: torch/autograd/gradcheck.py (hunk 4 above)
Comment (soulitzer) [reply torch/autograd/gradcheck.py:154]: Good catch, I shouldn't be doing this prematurely. In this PR v is still a python number. I'm moving the logic to handle tensors in the next PR.

File: torch/autograd/gradcheck.py (hunk 6 above)
Comment (soulitzer) [reply torch/autograd/gradcheck.py:265]: Looks a lot cleaner with this change!

File: torch/autograd/gradcheck.py (hunk 2 above)
Comment (soulitzer) [reply torch/autograd/gradcheck.py]: In the process of adding type hints I realized that `List[Optional[Tensor]]]` should be `List[Tensor]` in many places.

Comment (soulitzer) [reply torch/autograd/gradcheck.py]: In the previous PR, I added a test for the specific code path that tests this (where input is real, but we return both complex and real outputs).

File: torch/autograd/gradcheck.py [hunk 7]
```diff
@@ -141,11 +141,79 @@ def get_numerical_jacobian(fn, inputs, outputs=None, target=None, eps=1e-3,
...
+                # dL_dz_conj = 0.5 * [grad_out.conj() * ds_dx + grad_out * ds_dx.conj()]
+                #            = 0.5 * [grad_out.conj() * ds_dx + (grad_out.conj() * ds_dx).conj()]
+                #            = 0.5 * 2 * real(grad_out.conj() * ds_dx)
+                #            = real(grad_out.conj() * ds_dx)
+                jacobians_cols.append(torch.real(grad_out.conjugate() * ds_dx))
+            else:                   # R -> R
+                if isinstance(grad_out, complex):
+                    # placeholder if grad_out is complex but output is not
+                    jacobians_cols.append(torch.zeros_like(ds_dx))
+                    continue
```
Comment (soulitzer) [inline torch/autograd/gradcheck.py:202]: Previously we appended None here, but then I realized that line 212 wouldn't work properly (assigning NoneType to the column of a Tensor). The tests were still passing though because the code path was never reached, i.e., we didn't have any functions that take real input and have both complex and real output.

File: torch/autograd/gradcheck.py [hunk 8]
```diff
@@ -141,11 +141,79 @@ def get_numerical_jacobian(fn, inputs, outputs=None, target=None, eps=1e-3,
...
+    # Section 3.5.3 https://arxiv.org/pdf/1701.00392.pdf
+    # s = fn(z) where z = x for real valued input
+    # and z = x + yj for complex valued input
+    jacobians_cols: List[torch.Tensor] = []
+    ds_dx_tup = jvp_fn(delta)
+
+    if input_is_complex:            # C -> C, C -> R
+        ds_dy_tup = jvp_fn(delta * 1j)
+        for ds_dx, ds_dy in zip(ds_dx_tup, ds_dy_tup):
+            if not ds_dx.is_complex() and isinstance(grad_out, complex):
```
Comment (soulitzer) [inline torch/autograd/gradcheck.py:180]: Added this here as well, because C-> R case was missed previously.

Comment (zou3519) [reply torch/autograd/gradcheck.py:180]: Good catch!

**Review by zou3519 — APPROVED** [review]
lgtm, one suggestion
